File: agents/claude-code/cmd/claude-code/metadata.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/op-agent/opagent-protocol/go-sdk/op"
	"gopkg.in/yaml.v3"
)

type agentFrontMatter struct {
	Name        string `yaml:"name"`
	Description string `yaml:"description"`
}

func resolveAgentFile() (string, error) {
	if value := strings.TrimSpace(os.Getenv("OPAGENT_AGENT_FILE")); value != "" {
		if fileExists(value) {
			return value, nil
		}
		return "", fmt.Errorf("OPAGENT_AGENT_FILE does not exist: %s", value)
	}

	candidates := make([]string, 0, 8)
	if exe, err := os.Executable(); err == nil {
		exeDir := filepath.Dir(exe)
		candidates = append(candidates,
			filepath.Join(exeDir, "..", ".agent", "AGENT.md"),
			filepath.Join(exeDir, ".agent", "AGENT.md"),
			filepath.Join(exeDir, "..", "AGENT.md"),
		)
	}
	if cwd, err := os.Getwd(); err == nil {
		candidates = append(candidates,
			filepath.Join(cwd, ".agent", "AGENT.md"),
			filepath.Join(cwd, "AGENT.md"),
		)
		for dir := cwd; ; dir = filepath.Dir(dir) {
			candidates = append(candidates, filepath.Join(dir, "agents", agentID, "AGENT.md"))
			parent := filepath.Dir(dir)
			if parent == dir {
				break
			}
		}
	}

	for _, candidate := range candidates {
		cleaned := filepath.Clean(candidate)
		if fileExists(cleaned) {
			return cleaned, nil
		}
	}
	return "", fmt.Errorf("AGENT.md not found for %s", agentID)
}

func loadAgentMeta(agentFile string) (*op.AgentMeta, error) {
	raw, err := os.ReadFile(agentFile)
	if err != nil {
		return nil, fmt.Errorf("read agent metadata: %w", err)
	}
	frontMatter, err := parseAgentFrontMatter(string(raw))
	if err != nil {
		return nil, err
	}
	name := strings.TrimSpace(frontMatter.Name)
	if name == "" {
		name = "Claude Code"
	}
	description := strings.TrimSpace(frontMatter.Description)
	if description == "" {
		description = "Claude Code bridge for workspace-aware coding tasks."
	}
	return &op.AgentMeta{
		Name:        name,
		Description: description,
	}, nil
}

func loadAgentPrompt(agentFile string) (string, error) {
	raw, err := os.ReadFile(agentFile)
	if err != nil {
		return "", fmt.Errorf("read agent prompt: %w", err)
	}
	body, err := markdownBody(string(raw))
	if err != nil {
		return "", err
	}
	return expandPromptVariables(body), nil
}

func parseAgentFrontMatter(markdown string) (agentFrontMatter, error) {
	front, _, ok := splitFrontMatter(markdown)
	if !ok {
		return agentFrontMatter{}, nil
	}
	var parsed agentFrontMatter
	if err := yaml.Unmarshal([]byte(front), &parsed); err != nil {
		return agentFrontMatter{}, fmt.Errorf("parse agent front matter: %w", err)
	}
	return parsed, nil
}

func markdownBody(markdown string) (string, error) {
	_, body, ok := splitFrontMatter(markdown)
	if !ok {
		return strings.TrimSpace(markdown), nil
	}
	return strings.TrimSpace(body), nil
}

func splitFrontMatter(markdown string) (frontMatter string, body string, ok bool) {
	text := strings.TrimPrefix(strings.ReplaceAll(markdown, "\r\n", "\n"), "\ufeff")
	if !strings.HasPrefix(text, "---\n") {
		return "", markdown, false
	}
	remaining := text[len("---\n"):]
	idx := strings.Index(remaining, "\n---")
	if idx < 0 {
		return "", markdown, false
	}
	front := remaining[:idx]
	bodyStart := idx + len("\n---")
	if len(remaining) > bodyStart && remaining[bodyStart] == '\n' {
		bodyStart++
	}
	return strings.TrimSpace(front), remaining[bodyStart:], true
}

func expandPromptVariables(prompt string) string {
	replacements := map[string]string{
		"${platform}":  platformName(),
		"{{platform}}": platformName(),
	}
	out := prompt
	for placeholder, value := range replacements {
		out = strings.ReplaceAll(out, placeholder, value)
	}
	return out
}

func platformName() string {
	switch runtime.GOOS {
	case "darwin":
		return "macOS"
	case "windows":
		return "Windows"
	case "linux":
		return "Linux"
	default:
		return runtime.GOOS
	}
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

File: agents/claude-code/src/metadata.ts
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import type { AgentMeta } from "@op-agent/opagent-protocol";
import { platformName } from "./config.js";

const agentID = "claude-code";

export async function resolveAgentFile(env: NodeJS.ProcessEnv = process.env): Promise<string> {
  const explicit = env.OPAGENT_AGENT_FILE?.trim();
  if (explicit && await fileExists(explicit)) {
    return explicit;
  }

  const candidates: string[] = [];
  const moduleDir = path.dirname(fileURLToPath(import.meta.url));
  candidates.push(
    path.join(moduleDir, "..", "AGENT.md"),
    path.join(moduleDir, "..", "..", "AGENT.md"),
  );

  const cwd = process.cwd();
  candidates.push(
    path.join(cwd, ".agent", "AGENT.md"),
    path.join(cwd, "AGENT.md"),
  );
  for (let dir = cwd; ; dir = path.dirname(dir)) {
    candidates.push(path.join(dir, "agents", agentID, "AGENT.md"));
    const parent = path.dirname(dir);
    if (parent === dir) {
      break;
    }
  }

  for (const candidate of candidates) {
    const resolved = path.resolve(candidate);
    if (await fileExists(resolved)) {
      return resolved;
    }
  }
  throw new Error(`AGENT.md not found for ${agentID}`);
}

export async function loadAgentMeta(agentFile: string): Promise<AgentMeta> {
  const raw = await fs.readFile(agentFile, "utf8");
  const front = parseFrontMatter(raw);
  return {
    name: front.name || "Claude Code",
    description: front.description || "Claude Code bridge for workspace-aware coding tasks.",
  };
}

export async function loadAgentPrompt(agentFile: string): Promise<string> {
  const raw = await fs.readFile(agentFile, "utf8");
  return expandPromptVariables(markdownBody(raw));
}

export function markdownBody(markdown: string): string {
  const split = splitFrontMatter(markdown);
  return (split?.body ?? markdown).trim();
}

export function parseFrontMatter(markdown: string): { name?: string; description?: string } {
  const split = splitFrontMatter(markdown);
  if (!split) {
    return {};
  }
  const out: { name?: string; description?: string } = {};
  for (const line of split.frontMatter.split(/\r?\n/)) {
    const match = /^([A-Za-z0-9_-]+)\s*:\s*(.*)$/.exec(line);
    if (!match) {
      continue;
    }
    const key = match[1];
    const value = unquote(match[2].trim());
    if (key === "name") {
      out.name = value;
    } else if (key === "description") {
      out.description = value;
    }
  }
  return out;
}

function splitFrontMatter(markdown: string): { frontMatter: string; body: string } | null {
  const text = markdown.replace(/^\uFEFF/, "").replace(/\r\n/g, "\n");
  if (!text.startsWith("---\n")) {
    return null;
  }
  const remaining = text.slice("---\n".length);
  const idx = remaining.indexOf("\n---");
  if (idx < 0) {
    return null;
  }
  let bodyStart = idx + "\n---".length;
  if (remaining[bodyStart] === "\n") {
    bodyStart += 1;
  }
  return {
    frontMatter: remaining.slice(0, idx).trim(),
    body: remaining.slice(bodyStart),
  };
}

function expandPromptVariables(prompt: string): string {
  return prompt
    .replaceAll("${platform}", platformName())
    .replaceAll("{{platform}}", platformName());
}

function unquote(value: string): string {
  if ((value.startsWith("\"") && value.endsWith("\"")) || (value.startsWith("'") && value.endsWith("'"))) {
    return value.slice(1, -1);
  }
  return value;
}

async function fileExists(file: string): Promise<boolean> {
  try {
    const stat = await fs.stat(file);
    return stat.isFile();
  } catch {
    return false;
  }
}
